src/services/giant_api.rs

```diff
@@ -13,171 +14,222 @@ use crate::{
         uri::Uri,
     },
 };
-
-use clap::ValueEnum;
-
-use urlencoding::encode;
+use crate::model::blob::{Blob, BlobResp};
 
 #[derive(ValueEnum, Clone)]
 pub enum ListBlobsFilter {
     All,
     InMultiple,
 }
 
-pub fn get_client(giant_uri: &str) -> Result<Client, CliError> {
-    let auth_token = auth_store::get(giant_uri)?;
-    let mut headers = HeaderMap::new();
-    headers.insert("Authorization", auth_token.parse()?);
-
-    Ok(Client::builder().default_headers(headers).build()?)
+pub struct GiantApiClient {
+    client: Client,
+    base_url: Url,
 }
 
-pub fn check_hash_exists(uri: &str, hash: &str) -> Result<bool, CliError> {
-    let mut url = String::from(uri);
-    url.push_str("/api/resources/");
-    url.push_str(hash);
-    url.push_str("?basic=true");
-
-    let client = get_client(uri)?;
+impl GiantApiClient {
+    pub fn new(base_url: Url) -> Self {
+        Self {
+            client: {
+                let auth_token = auth_store::get(base_url.as_str()).unwrap();
+                let mut headers = HeaderMap::new();
+                headers.insert("Authorization", auth_token.parse().unwrap());
+                Client::builder().default_headers(headers).build().unwrap()
+            },
+            base_url,
+        }
```

Comment:
A few notes on this...

I would lift this logic up immediately after the opening brackets in order to keep the return as clear as possible, something like this:

```rust
    pub fn new(base_url: Url) -> Self {
        let auth_token = auth_store::get(base_url.as_str()).unwrap();
        let mut headers = HeaderMap::new();
        let client = Client::builder().default_headers(headers).build().unwrap();
        headers.insert("Authorization", auth_token.parse().unwrap());
        Self { client, base_url }
    }
```

It's also probably worth replacing those `unwrap`s with `?` - but this is a bit more convoluted because it requires to change the return type of `new()` as well as `main()` in `main.rs`, and you'd also have to add `?`s at every call site.

```rust
    pub fn new(base_url: Url) -> anyhow::Result<Self> {
        let auth_token = auth_store::get(base_url.as_str())?;
        let mut headers = HeaderMap::new();
        headers.insert("Authorization", auth_token.parse()?);
        let client = Client::builder().default_headers(headers).build()?;
        Ok(Self { client, base_url })
    }
```

Reply:
Yep will make the return tweak. I'm fine with `unwrap()` for now since it's a CLI and we can't reasonably recover from this. Later on I might refactor to propagate a `Result` up to the top level.